### PR TITLE
OpenColorIO: Fix validation & package_info

### PR DESCRIPTION
**Pull request by irieger**
### Summary
Changes to recipe:  **OpenColorIO/all**

#### Motivation
Ensure more configurations options are properly build. The validation basically prevents all shared builds right now.

#### Details
* Remove the validation preventing shared building: I synced with the maintainer a bit in the projects Slack channel and he definitely confirmed, that there is no intention from the project side that things should only work with static linking
* Add required libraries to package_info: Not sure if that is required, but most projects seem to list all dependencies

---
- [x] Read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md)
- [x] Checked that this PR is not a duplicate: [list of PRs by recipe](https://github.com/conan-io/conan-center-index/discussions/24240)
- [ ] Tested locally with at least one configuration using a recent version of Conan

### CODE REVIEW

**Comment by irieger** [comment]
Taking out the relevant parts of #23112 without the build hacks done there by changing linking to public.

**Comment by RubenRBS** [comment]
Let me know if you need help fixing the gcc compilation issue/ping me once you figure it out, thanks! :)

**Comment by irieger** [comment]
> Let me know if you need help fixing the gcc compilation issue/ping me once you figure it out, thanks! :)

Hey @RubenRBS , thanks for offering help. Just wanted to have a more clean MR as documentation compared to the old one, but the underlying problem is discussed in #23421 and still is true. Looks like it works with gold, I tried mold and kept it for now in my profile and it works well so far. I have no clue what the proper fix for this might be. It just looks like the default linker is requireing to know where to find each indirectly needed so while mold and gold are fine with linking just the direct dependencies. (Populating the LD_LIBRARY_PATH before build makes default linking work too.)

File: recipes/opencolorio/all/conanfile.py [hunk 1]
```diff
@@ -189,6 +186,20 @@ def package_info(self):
 
         self.cpp_info.libs = ["OpenColorIO"]
 
+        self.cpp_info.requires.append("expat::expat")
```
Comment (AbrilRBS) [inline recipes/opencolorio/all/conanfile.py:189]: These requirements are only necesary when some of them are optional, or where not the whole package is needed. If no requirements are specified, all available ones coming from the direct dependencies are used

So I'd say we can revert this bit, as that's what this code is doing, adding all available ones. We can let Conan handle it in this case :)

**Review by AbrilRBS — COMMENTED** [review]
Thanks! Comment a about the requires

File: recipes/opencolorio/all/patches/2.3.1-0001-fix-cmake-source-dir-and-targets.patch [hunk 2]
```diff
@@ -68,7 +68,7 @@ index 26b4bb4c..d54cff3d 100755
          "$<BUILD_INTERFACE:xxHash>"
 -        ${YAML_CPP_LIBRARIES}
 -        MINIZIP::minizip-ng
-+        yaml-cpp
++        PUBLIC yaml-cpp
```
Comment (memsharded) [inline recipes/opencolorio/all/patches/2.3.1-0001-fix-cmake-source-dir-and-targets.patch:71]: I am bringing this back, as it is related to a limitation of CMakeDeps, sorry for the noise.
There is already some work in a new improved ``CMakeDeps`` generator that will make this unnecessary.

**Review by AbrilRBS — APPROVED** [review]

**Review by memsharded — APPROVED** [review]

**Pull request merged**